napari/layers/points/_points_key_bindings.py

```diff
@@ -58,7 +60,61 @@ def paste(layer: Points):
     trans._("Select all points in the current view slice."),
 )
 def select_all(layer: Points):
-    layer.selected_data = set(layer._indices_view[: len(layer._view_data)])
+    new_selected = set(layer._indices_view[: len(layer._view_data)])
+
+    # If all visible points are already selected, deselect the visible points
+    if new_selected & layer.selected_data == new_selected:
+        layer.selected_data = layer.selected_data - new_selected
+        show_info(
+            trans._(
+                "Deselected all points in this slice,"
+                " use Shift-A to deselect all points on the layer."
+                f" ({len(layer.selected_data)} selected)"
+            )
+        )
+
+    # If not all visible points are already selected, additionally select the visible points
+    else:
+        layer.selected_data = layer.selected_data | new_selected
+        show_info(
+            trans._(
+                f"Selected {len(new_selected)} points in this slice,"
+                " use Shift-A to select all points on the layer."
+                f" ({len(layer.selected_data)} selected)"
+            )
+        )
+    layer._set_highlight()
+
+
+@register_points_action(
+    trans._("Select all points in the layer."),
+)
+def select_all_3d(layer: Points):
+    new_selected = set(range(layer.data.shape[0]))
+    # Needed for the notification
+    view_selected = set(layer._indices_view[: len(layer._view_data)])
+
+    # If all points are already selected, deselect all points
+    if layer.selected_data == new_selected:
```

Comment:
This can be sped up because there cannot be any selected data that is not in data, so you can just do `if len(layer.selected_data) == len(data):` — and indeed you can do this *before* creating `new_selected`, saving even more time.

For the notification, you can just say "Cleared all selections" — no need to specify in-slice/out-of-slice, 0 is always 0. 😃

Reply:
True :D Done